### PR TITLE
Writing Flow/Rich Text: unify split logic

### LINKED ISSUE #53076: Paragraph: paste contents at cursor point rather than as new blocks
Hat tip @richtabor for the text and video: When selecting, copying, and pasting, across more than one paragraph block, the selected contents are not pasted at the cursor point. Unexpectedly, the contents are pasted as new blocks. 

The expectation is that one pastes the copied contents of more than one block, right where the cursor is. The contents should inherit the attributes of the pasted-in paragraph. 

https://github.com/WordPress/gutenberg/assets/26996883/6f5b7d5b-b2f8-424c-b82b-9e0d21b662b6

cc @ellatrix for consideration as you weave your way through writing flow issues.

### PR DESCRIPTION
**Pull request by ellatrix**
## What/why?

Fixes https://github.com/WordPress/gutenberg/issues/53076: when pasting blocks, the split block should be merged with the pasted blocks.

This PR aims to unify split logic (enter key and paste). Currently we have logic _both_ in the block editor generally and rich text specifically when it contains selection. This PR moves all the logic to the block editor (writing flow), adding a block.json setting that tells us if we can split a block.

This removes the need for the block to pass an `onSplit` function, the behaviour can be turned on with just a setting. This is also quite convenient for extenders wishing to recreate the core blocks.

This will also allow us to solve splitting for paste more easily, which is actually a split operation + merging the ends together.

## How?

One implementation details is that change of how rich text listens for key events. Right now in trunk, the event listener is attached to the rich text node, so writing flow (listening on a wrapper node) cannot prevent default behaviour. To fix this, rich text can listen to the document instead. This is how it works for browser behaviour too. Default behaviour is cancelable until the event bubbled all the way through the DOM. Rich text is meant to be a sort of polyfill for the native `contenteditable` API: actions should be cancelable.

## Testing Instructions

Copy a few paragraphs and paste them into another paragraph. The result should be:

```
[target paragraph][pasted paragraph

pasted paragraph

pasted paragraph][target paragraph]
```

## Screenshots or screencast 

paste-between

### CODE REVIEW

File: packages/block-editor/src/store/actions.js [hunk 1]
```diff
@@ -929,10 +938,62 @@ export const __unstableSplitSelection =
...
+			! blocks.length &&
+			selectionA.clientId === selectionB.clientId &&
+			selectionA.attributeKey === selectionB.attributeKey &&
+			selectionA.offset === selectionB.offset &&
+			! select.getBlock( selectionA.clientId ).innerBlocks?.length
+		) {
+			function getRichTextAttributeLength( clientId, attributeKey ) {
+				return create( {
+					html: select.getBlockAttributes( clientId )[ attributeKey ],
+				} ).text.length;
```
Comment (ellatrix) [inline packages/block-editor/src/store/actions.js]: If the rich text value were in the store, we could simply check `.length`...

**Review by richtabor — APPROVED** [review]
UX/writing flow wise, this feels solid, with more expected results on copying/pasting across multiple blocks.

**Review by Mamaduka — COMMENTED** [review]
Maybe we should leave the `onSplit` callback if the block has more specific needs for this logic.

A good example would be the button block. Currently, it copies all attributes on the split action, including URLs. Which I think is a bug; only styling attributes and text included on the split action should be copied.

**Comment by ellatrix** [comment]
@Mamaduka Right, that's a good point. I don't like keeping onSplit though, maybe you should be able to opt-out of attributes in the block supports key? Or opt-in? What do you think?

```
splitting: {
  url: false
}
```

Or maybe a transform?

**Comment by Mamaduka** [comment]
@ellatrix, I thought we'd have to support the legacy method for backward compatibility.

**Comment by ellatrix** [comment]
@Mamaduka I meant, how to incorporate it in the new way?

**Comment by ellatrix** [comment]
> A good example would be the button block. Currently, it copies all attributes on the split action, including URLs. Which I think is a bug; only styling attributes and text included on the split action should be copied.

Thinking more about this, I don't think this is a bug. It you split a button in the middel, it will retain all properties for both blocks, including URLs. I think that's expected. If you split at the start or end, a new button will be created at the start or end and no URL is copied over.

**Comment by ellatrix** [comment]
Feeling great about this change. Let's merge and iterate, it's good to have some testing in the GB before we have this in WP.

**Comment by Mamaduka** [comment]
> Thinking more about this, I don't think this is a bug. It you split a button in the middel, it will retain all properties for both blocks, including URLs. I think that's expected. If you split at the start or end, a new button will be created at the start or end and no URL is copied over.

Makes sense.

**Comment by t-hamano** [comment]
It seems that due to this PR, when you paste text into a text field inside a block, it is always converted to a Paragraph block.

See #61377

**Comment by Mamaduka** [comment]
Good catch, @t-hamano!

It's odd that the e2e tests missed this; we have an extensive suite for the Embed block. When the issue is fixed, we should probably introduce a new testxed, maybe using the RSS block.

File: packages/rich-text/src/component/event-listeners/copy-handler.js [hunk 2]
```diff
@@ -30,10 +30,12 @@ export default ( props ) => ( element ) => {
 		}
 	}
 
-	element.addEventListener( 'copy', onCopy );
-	element.addEventListener( 'cut', onCopy );
+	const { defaultView } = element.ownerDocument;
+
+	defaultView.addEventListener( 'copy', onCopy );
+	defaultView.addEventListener( 'cut', onCopy );
```
Comment (Mamaduka) [inline packages/rich-text/src/component/event-listeners/copy-handler.js:36]: I believe this change produces a `TypeError: Cannot read properties of null (reading 'addEventListener')` error. I'm having trouble reproducing it manually, but it's consistent when running e2e tests.

```
npm run test:e2e -- -g "can be inserted after refresh"
```

## Screenshot
CleanShot 2024-05-07 at 16 27 41

**Comment by matthewmcvickar** [comment]
This comment is a followup and context for [my comment on the WP Dev Blog post that mentions this change](https://make.wordpress.org/core/2024/06/24/miscellaneous-editor-changes-in-wordpress-6-6/#comment-46863), where I expressed confusion about this logic and its settings in `block.json`.

My comment, for convenience:

> I’m having trouble understanding the `splitting` attribute and what the `true` and `false` values mean for editor behavior.
> 
> Let’s say I have a custom block that’s just a `<p>` tag (its own block—not a block variation—because I want to disable typography and colors for that block and set a particular class on the tag). If I set `splitting` to `true`, pressing the <kbd>Return</kbd> key in the block does nothing, and trying to paste content that has line returns it also does nothing. (Pasting content without line returns is allowed.) If I set `splitting` to false, I can do both of these things. This seems counterintuitive or at least under-explained. Is there documentation about this that I’m missing?
> 
> Also: previously I had used the `onSplit` prop for handling the <kbd>Return</kbd> key in a block. For some blocks, I want hitting the <kbd>Return</kbd> key to exit that block and start a new empty paragraph—how should I accomplish that? Being 'trapped' in the block and having the <kbd>Return</kbd> key do nothing is not desirable or expected behavior, given that, e.g., the Heading block allows for 'escaping' its block when hitting <kbd>Return</kbd>.

Seeing this comment, @ellatrix asked to see my code to help clarify. Here is my block code:

**`block.json`**

```json
{
	"$schema": "https://schemas.wp.org/trunk/block.json",
	"apiVersion": 3,
	"name": "baz-qux/page-tagline",
	"title": "Page Tagline",
	"category": "baz-qux",
	"icon": "megaphone",
	"description": "A short introduction to the page.",
	"keywords": ["intro", "excerpt", "description"],
	"supports": {
		"anchor": false,
		"align": false,
		"splitting": false
	},
	"editorScript": "file:./page-tagline.js"
}
```

**`page-tagline.jsx`** (compiles into `page-tagline.js`)

```jsx
import metadata from './block.json';

const { registerBlockType } = wp.blocks;
const { useBlockProps, RichText } = wp.blockEditor;

registerBlockType( metadata.name, {
	attributes: {
		content: {
			type: 'string',
			source: 'text',
			selector: 'p'
		}
	},
	edit: (props) => {
		return <RichText
			{ ...useBlockProps() }
			tagName="p"
			className="wp-block wp-block-paragraph page-tagline"
			value={ props.attributes.content }
			placeholder="Add tagline"
			onChange={ ( newContent ) => props.setAttributes({ content: newContent }) }
		/>;
	},
	save: (props) => {
		return <RichText.Content
			{ ...useBlockProps.save() }
			tagName="p"
			className="page-tagline"
			value={ props.attributes.content }
		/>;
	}
} );
```

Thank you for your help!

**Comment by ellatrix** [comment]
Hi! I tried your code, and for me it looks like `splitting: false` works correctly. It will not split the block into two blocks but rather insert a soft line break.

splitting-false

Now, you are right that setting `splitting` to true will do nothing. That's because `RichText` in the edit function is missing the `identifier` prop:

```js
<RichText
	{ ...useBlockProps() }
	identifier="content"
```

This should allow the block to be split into two blocks. We should indeed mention that in documentation.

It's also worth noting that if you'd like to allow _merging_ these blocks back together, you should provide a `merge` function and pass down `mergeBlocks` to RichText as `onMerge`:

```js
merge( attributes, attributesToMerge ) {
	return {
		content:
			( attributes.content || '' ) +
			( attributesToMerge.content || '' ),
	};
},
edit: function Edit( props ) {
	return (
		<RichText
			{ ...useBlockProps() }
			identifier="content"
			onMerge={ props.mergeBlocks }
```

**Comment by ellatrix** [comment]
I've added docs here: #63016

**Comment by matthewmcvickar** [comment]
Thank you for the followup! This is helpful. I am now getting the behavior I expect, and it's nice that hitting <kbd>Enter</kbd> at the end of a block that has `splitting` set to `true` creates a new `core/paragraph` block by default.

(Am I correct in assuming that the `merge()` function in the `registerBlockType()` arguments is a shortcut to define the `mergeBlocks` prop? That `registerBlockType( name, { merge(...) {...} } )` is equal to `registerBlockType( name, { mergeBlocks: function(...) {...} } )` in the same way that `edit:` or `edit()` can be used?)

One more thing: I'm unclear from the discussion on this ticket whether `onSplit` will be sticking around. It's still being used on some core blocks, right? How will custom splitting behavior be supported if the `onSplit` prop is deprecated?

**Comment by ellatrix** [comment]
As far as I know, there's no `mergeBlocks` property for block registration. There _is_ a `mergeBlocks` prop given to the `edit` function that you need to pass down to RichText as `onMerge`, as illustrated above. Otherwise it won't work. Note that these APIs are a bit in need of a change, similar to splitting.

`onSplit` is officially deprecated but will probably keep working as it is now for many years, if not forever. Until it is a burden to maintain, but right now it is not.

**Comment by t-hamano** [comment]
Another regression: It appears that it is no longer possible to paste text from _outside_ WordPress into non-modal classic block. See #64771.

**Comment by elzadj** [comment]
@ellatrix Can we please have some input here? https://github.com/WordPress/gutenberg/issues/65274#issuecomment-2353289668
Thanks

**Comment by keepyourreceipt** [comment]
Hi everyone. Theme / plugin developer here. Thanks for all the work you do on this project :)

I have a question, from what I can tell this might be the best place to post it. 

With the changes to how onSplit works, I don't have access to the attributes of the new block that's created on split (as I understand). Could we add an option to the splitting block.json setting to have the new block copy the attribute values of the one being copied? So new block has all the same attribute values of old block. 

@ellatrix your comment here https://github.com/WordPress/gutenberg/pull/54543#issuecomment-2092751159 seems to reference this, but I wasn't sure if I fully understood.

Usage: I have a custom paragraph block with an alignment attribute. With the new split behavior, if the original block is set to center align, the new block comes in with the align set to left, the block's default value. The author has set the alignment to "center" each time they add a new paragraph. Kind of annoying. Looks like I could have resolved this with the old implementation.

Screenshot 2024-10-13 at 2 39 07 PM

Alternatively, is there a way to set default attributes on the new block added on split? I've been searching around, but haven't found any good info on this. Hence my post here.

Thanks.

**Comment by ellatrix** [comment]
> Usage: I have a custom paragraph block with an alignment attribute. With the new split behavior, if the original block is set to center align, the new block comes in with the align set to left, the block's default value. The author has set the alignment to "center" each time they add a new paragraph. Kind of annoying. Looks like I could have resolved this with the old implementation.

Hm, when I split a core paragraph block, it keeps the alignment. Even when I split a core heading block with alignment, it keeps it on the newly created paragraph. Could you share some minimal block code that triggers the problem?

**Comment by keepyourreceipt** [comment]
Hey @ellatrix thanks for the reply :)

I've included the block.json and edit function below. With this code, on split, the attribute "align" is set to the default value of "left" when the new block is added to the editor.

Please let me know if there's any additional code or examples I can provide.

Thanks.

**block.json**
```json
{
    "apiVersion": 3,
    "name": "thirtysix-blocks/paragraph",
    "title": "Paragraph",
    "description": "Displays a rich text editor.",
    "category": "layout",    
    "textdomain": "thirtysix-blocks",
    "editorScript": "file:./index.js",
    "viewScript": "file:./view.js",
    "style": "file:./style-index.css",
    "editorStyle": "file:./index.css",
    "viewStyle": "file:./view.css",
    "ancestor": [
        "thirtysix-blocks/text",
        "thirtysix-blocks/banner",
        "thirtysix-blocks/column",
        "thirtysix-blocks/posts"
    ],
    "attributes": {
        "text": {
            "type": "string"            
        },        
        "align": {
            "type": "string",
            "default": "left"
        },        
        "client_id": {
            "type": "string"
        }                      
    },
    "supports": {
        "color": {
	    "text": true,	
            "background": false		
	},
        "typography": {
            "fontSize": true
        },
        "renaming": false,
        "customClassName": false,
        "splitting": true
    }
}
```

**edit.js**
```js
import "./editor.scss";
import "./style.scss";

import { useEffect } from "@wordpress/element";
import { useBlockProps, RichText } from "@wordpress/block-editor";

import ParagraphInspectorControls from "./_inspector-controls";
import ParagraphBlockControls from "./_block-controls";

export default function Edit({ attributes, setAttributes, clientId }) {
  const blockProps = useBlockProps({
    style: {
      color: attributes.color,
      textAlign: attributes.align,
    },
  });

  useEffect(() => {
    setAttributes({ client_id: clientId });
  }, [clientId]);

  return (
    <>
      <ParagraphInspectorControls
        attributes={attributes}
        setAttributes={setAttributes}
      />

      <ParagraphBlockControls
        attributes={attributes}
        setAttributes={setAttributes}
      />

      <section {...blockProps}>
        <RichText
          value={attributes.text}          
          allowedFormats={[
            "core/text-color",
            "core/bold",
            "core/italic",
            "core/link",
            "core/superscript",
            "core/subscript",
          ]}
          tagName="p"          
          identifier={"text"}  
          placeholder="Lorem ipsum dolor sit amet, consectetur adipiscing elit..."
          onChange={(value) => setAttributes({ text: value })}                    
        />
      </section>
    </>
  );
}
```

**Comment by ellatrix** [comment]
Hi @keepyourreceipt!

Unfortunately I will need more information, because I don't see an alignment button. Here's the code you provided that is executable in the console:

```js
wp.blocks.registerBlockType('thirtysix-blocks/paragraph', {
	apiVersion: 3,
	name: "thirtysix-blocks/paragraph",
	title: "Paragraph",
	description: "Displays a rich text editor.",
	category: "layout",
	textdomain: "thirtysix-blocks",
	attributes: {
		text: {
			type: "string"
		},
		align: {
			type: "string",
			default: "left"
		},
		client_id: {
			type: "string"
		}
	},
	supports: {
		color: {
			text: true,
			background: false
		},
		typography: {
			fontSize: true
		},
		renaming: false,
		customClassName: false,
		splitting: true
	},
	edit: function Edit({ attributes, setAttributes, clientId }) {
		const blockProps = wp.blockEditor.useBlockProps({
			style: {
				color: attributes.color,
				textAlign: attributes.align,
			},
		});

		wp.element.useEffect(() => {
			setAttributes({ client_id: clientId });
		}, [clientId]);

		return wp.element.createElement(
			'section',
			blockProps,
			wp.element.createElement(wp.blockEditor.RichText, {
				value: attributes.text,
				allowedFormats: [
					"core/text-color",
					"core/bold",
					"core/italic",
					"core/link",
					"core/superscript",
					"core/subscript",
				],
				tagName: "p",
				identifier: "text",
				placeholder: "Lorem ipsum dolor sit amet, consectetur adipiscing elit...",
				onChange: (value) => setAttributes({ text: value })
			})
		);
	}
});
```

Could you add your alignment functionality to this?

**Pull request merged**